discovery-provider/src/tasks/index_aggregate_plays.py

```diff
@@ -1,17 +1,103 @@
 import logging
 import time
+import sqlalchemy as sa
 from src.tasks.celery_app import celery
+from src.models import LatestSlots, Play
 
 logger = logging.getLogger(__name__)
 
+AGGREGATE_PLAYS_TABLE_NAME = "aggregate_plays"
 
-def update(self, db):
-    with db.scoped_session() as session:
-        start_time = time.time()
-        session.execute("REFRESH MATERIALIZED VIEW CONCURRENTLY aggregate_plays")
+### UPDATE_AGGREGATE_PLAYS_QUERY ###
+# Get new plays that came after the previous most recent indexed slot for aggregate_play
+# Group those new plays by play item id to get the aggregate play counts
+# For new play item ids, insert those aggregate counts
+# For existing play item ids, add the new aggregate count to the existing aggregate count
+UPDATE_AGGREGATE_PLAYS_QUERY = """
+    WITH aggregate_plays_latest_slot AS (
+        SELECT
+            :prev_latest_slot AS prev_latest_slot,
+            :new_latest_slot AS new_latest_slot
+    ),
+    new_plays AS (
+        SELECT
+            play_item_id,
+            count(play_item_id) AS count
+        FROM
+            plays p
+        WHERE
+            p.slot > (
+                SELECT
+                    prev_latest_slot
+                FROM
+                    aggregate_plays_latest_slot
+            )
+            AND p.slot <= (
+                SELECT
+                    new_latest_slot
+                FROM
+                    aggregate_plays_latest_slot
+            )
+        GROUP BY
+            play_item_id
+    )
+    INSERT INTO
+        aggregate_plays (play_item_id, count)
+    SELECT
+        new_plays.play_item_id,
+        new_plays.count
+    FROM
+        new_plays ON CONFLICT (play_item_id) DO
+    UPDATE
+    SET
+        count = aggregate_plays.count + EXCLUDED.count
+    """
+
+UPSERT_LATEST_SLOTS = """
+    INSERT INTO latest_slots (tablename, slot)
+    VALUES(:tablename, :slot)
+    ON CONFLICT (tablename)
+    DO UPDATE SET slot = EXCLUDED.slot;
+    """
+
+def _update_aggregate_plays(session):
+    # get the last updated slot that counted towards the current aggregate plays
```

Comment:
Since there can be multiple plays in a slot, I'm a little concerned about aggregating a slot that only has some of its plays indexed. I confirmed with Hareesh that we shouldn't happen. We still may want to use play IDs instead.

Reply:
couldn't we just do slot >= prev_latest_slot and recalculate and update the value for the prev_latest_slot to ensure we didn't miss anything?

Reply:
I added the upper bound to avoid counting the same slot twice. If there was only the lower bound, I think there's a risk that plays would get new rows and the new latest_slot value would be lower than the actual window used to calculate the aggregates.

Reply:
i was talking about making the lower bound condition in the query >= from > like it is currently. if i understand correctly, the problem is that if you run this aggregation in the middle of play indexing while it's inserting records for a slot you could miss some of the most recent records. my suggestion with moving lower bound to >=, it queries all plays for that slot again and if the value already exists in aggregate_plays it overwrites it with the new, updated value.

Reply:
update - can't do this since we're adding not recalculating the value

Reply:
There was some confusion, discussed offline. We're now considering if it's safe to include the latest slot in an aggregate. Asking @hareeshnagaraj  to make sure this logic is ok with how solana plays are indexed. From my understanding, aggregating by slot number like this is ok as long as there are no gaps or partial slots. If there were any gaps in sol plays, indexing sol plays would be broken. If the plays table ever has an incomplete slot with some plays missing, then there's a risk of missing plays in the aggregate. It's important to not miss since this sliding window counts each play exactly once, no re-calculating because we want to be able to prune old plays.

Reply:
Worth noting that these changes make it impossible for us to shut off solana play indexing going forward. To your points @isaacsolo, im starting to think that using the play ids like you're calling out makes more sense. What are the drawbacks to that approach? 

Even if there is some non-determinism in how plays are indexed from solana & otherwise into the plays table, the ids will make sure we don't over count and get exactly what we want

Reply:
Could you explain why we can shut off solana play indexing?

Reply:
If we stopped producing plays with slots in them, they wouldn't trigger updates to the aggregate view b/c the latest slot wouldn't move. Eventually when we turn solana plays back on, then we probably would get the updates, but it might be a huge amount of them

Reply:
this particular portion of the population seems fine, i might suggest running some amount of slots behind intentionally so we aren't pushing right up against the tail of the chain - otherwise this seems okay, had some other comments in addition to ray's about the structure of LatestIndexingStates though

Reply:
I hadn't considered shutting off solana play indexing. The only strange thing about using id instead of slot is that nothing else depends on it. I think the only reason we have it is to keep some primary key on this table as a good practice. Fortunately, that's what we need in this case. As long as there isn't anything we're missing from slots, I think we should just use the primary key id.

Reply:
agree. going forward, i think we should ensure all tables get an auto-assign auto-increment primary key. they always end up useful, so you don't have to depend on ROW NUMBER or whatever